### PR TITLE
[done?] SerializableTransactionRunner onFailure/onSuccess callback

### PR DESCRIPTION
**Pull request by TheRealMarnes**
Inspired by something I wanted to do in my private project: log some info when SerializableTransactionRunner encounters a serialization failure, and a final time upon success.

> transaction failed 1 time! retrying...
> transaction failed 2 times! retrying...
> transaction succeeded after 2 retries!

### CODE REVIEW

File: core/src/main/java/org/jdbi/v3/core/transaction/SerializableTransactionRunner.java [hunk 1]
```diff
@@ -55,16 +61,16 @@ public SerializableTransactionRunner(TransactionHandler delegate) {
...
-                    stack.addSuppressed(last);
-                }
+                stack.push(x);
+                config.onFailure.accept(new ArrayList<>(stack));
 
                 // no more attempts left? Throw ALL the exceptions! \o/
                 if (--attempts <= 0) {
-                    throw stack;
+                    X toThrow = stack.removeFirst();
+                    while (!stack.isEmpty()) {
```
Comment (stevenschlansker) [inline core/src/main/java/org/jdbi/v3/core/transaction/SerializableTransactionRunner.java]: `stack.stream().forEach(toThrow::addSuppressed);`

File: RELEASE_NOTES [hunk 2]
```diff
@@ -1,3 +1,7 @@
+3.2.1
```
Comment (stevenschlansker) [inline RELEASE_NOTES]: this is new api so it needs to be a new minor: `3.3.0`

Comment (TheRealMarnes) [reply RELEASE_NOTES]: k

**Comment by TheRealMarnes** [comment]
Not that it really matters, but should the _first_ or _last_ exception be thrown with the others suppressed?

**Comment by TheRealMarnes** [comment]
@stevenschlansker 

image

❤️ jdbi :smiley:

image

**Pull request merged**